pkg/storage/fs/cephfs/utils.go

```diff
@@ -0,0 +1,197 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+//+build ceph
+
+package cephfs
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"github.com/rs/zerolog"
+	"io"
+	"path/filepath"
+	"runtime"
+	"strconv"
+	"strings"
+
+	cephfs2 "github.com/ceph/go-ceph/cephfs"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+)
+
+// Mount type
+type Mount = *cephfs2.MountInfo
+
+// Statx type
+type Statx = *cephfs2.CephStatx
+
+var dirPermFull = uint32(0777)
+var dirPermDefault = uint32(0775)
+var filePermDefault = uint32(0660)
+
+func Logme(log *zerolog.Logger) {
+	_, _f, _n, _ := runtime.Caller(1)
+	log.Debug().Msgf("[debug] %s:%d", _f, _n-1)
+}
+
+/*
+func firstEntry(dir *cephfs2.Directory) *cephfs2.DirEntry {
+	for dirEntry, err := dir.ReadDir(); dirEntry != nil; dirEntry, err = dir.ReadDir() {
+		if err != nil {
+			return nil
+		}
+		if strings.HasPrefix(dirEntry.Name(), ".") {
+			continue
+		}
+		return dirEntry
+	}
+
+	return nil
+}
+*/
+
+func closeDir(directory *cephfs2.Directory) {
+	if directory != nil {
+		_ = directory.Close()
+	}
+}
+
+func closeFile(file *cephfs2.File) {
+	if file != nil {
+		_ = file.Close()
+	}
+}
+
+func destroyCephConn(mt Mount, perm *cephfs2.UserPerm) *cacheVal {
+	if perm != nil {
+		perm.Destroy()
+	}
+	if mt != nil {
+		_ = mt.Release()
+	}
+	return nil
+}
+
+func deleteFile(mount *cephfs2.MountInfo, path string) {
+	_ = mount.Unlink(path)
+}
+
+/*
+func dirCmp(currPath string, curr *cephfs2.Directory, prev *cephfs2.Directory) (ri []*provider.RecycleItem, err error) {
+	cEntry := firstEntry(curr)
+	pEntry := firstEntry(prev)
+
+	for true {
+		if cEntry.Name() < pEntry.Name() {
+			cEntry, err = curr.ReadDir(); if err != nil { return }
+		} else if cEntry.Name() == pEntry.Name() {
+			if cEntry.DType() == cephfs2.DTypeDir && cEntry
+		}
+	}
+
+	return
+}
+*/
+
+func isDir(t provider.ResourceType) bool {
+	return t == provider.ResourceType_RESOURCE_TYPE_CONTAINER
+}
+
+func (fs *cephfs) makeFIDPath(fid string) string {
+	return filepath.Join(fs.conf.FIDFolder, fid)
+}
+
+func makeFID(user *User, t provider.ResourceType, absolutePath string, stat Statx) (rid *provider.ResourceId, err error) {
+	if t == provider.ResourceType_RESOURCE_TYPE_SYMLINK {
+		return &provider.ResourceId{OpaqueId: ""}, nil
+	}
+
+	hash := md5.New()
+	hash.Write([]byte(absolutePath))
+	fid := hex.EncodeToString(hash.Sum(nil)) + "_" + strconv.FormatUint(uint64(stat.Inode), 10)
```

Comment:
you should write a uuid into the extended attributes and use that as the name of the symlink. That way you can rebuild the fileid to path mapping in case it was not maintained by a process like inotify. When bypassing reva this allows reva to detect new files (that don't have a uuid, yet) and update the fileid to path mapping on the fly.

Furthermore, the link target should not contain the absolute path: it makes renames expensive because you have to traverse all children and update their symlink target as well. Instead use a relative symlink in the form `../<parentid>/<name>`. This reduces the number of additional writes to the rename of of the actual folder to one: the changing of the symlink target of the renamed fileid. The tradeoff is that looking up the path now requires reading multiple links and building the path by joining the names, but they are likely already cached because navigation happens from the root.

Reply:
I can write a UUID+inode on the xattrs instead. I already write on line 136 the id on the file which describes the name of the symlink.

Yeah, If a directory changes name, then you would need to update the symlinks of all its children. I can do the `../parentid/name` thing, or I can ditch subvolumes and use hardlinks, so I don't get `Invalid cross-device link` errors...
But that won't support Openstack Manila, but there may be a workaround.